File: supabase/functions/_shared/golden-athlete-timelines_test.ts
import { assert, assertEquals } from 'jsr:@std/assert@1';
import { evaluateAdaptiveSafety } from './adaptive-safety.ts';

type Json = Record<string, unknown>;

const fixture = JSON.parse(
  await Deno.readTextFile(
    new URL(
      '../../../test/fixtures/golden_athlete_timelines.json',
      import.meta.url,
    ),
  ),
) as { timelines: Json[] };

Deno.test('server safety replays every golden athlete day', () => {
  for (const timeline of fixture.timelines) {
    for (const day of timeline.days as Json[]) {
      const evidence = day.evidence as Json;
      const expected = day.expected as Json;
      const result = evaluateAdaptiveSafety({
        illness: evidence.illness_symptoms === true,
        injury: evidence.injury_pain === true,
        athleteState: evidence.illness_symptoms === true ||
            evidence.injury_pain === true
          ? 'RED'
          : Number(evidence.readiness) < 45
          ? 'ORANGE'
          : Number(evidence.readiness) < 60
          ? 'YELLOW'
          : 'GREEN',
        availableMinutes: Number(evidence.available_minutes),
      });
      const safetyRest = evidence.illness_symptoms === true ||
        evidence.injury_pain === true || Number(evidence.available_minutes) <= 0;
      assertEquals(
        result.outcome,
        safetyRest ? 'REST' : 'CONTINUE',
        `${timeline.id} ${day.date}`,
      );
      if (result.outcome === 'REST') {
        assert(
          (expected.allowed_actions as string[]).includes('REST'),
          `${timeline.id} ${day.date}`,
        );
      }
    }
  }
});

File: test/golden_athlete_timelines_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';

void main() {
  final fixture = jsonDecode(
    File('test/fixtures/golden_athlete_timelines.json').readAsStringSync(),
  ) as Map<String, dynamic>;
  final timelines = (fixture['timelines'] as List).cast<Map<String, dynamic>>();

  test('covers every required anonymized athlete scenario', () {
    expect(fixture['schema_version'], '1.0');
    expect(
      timelines.map((timeline) => timeline['id']).toSet(),
      {
        'normal_training',
        'overload',
        'illness',
        'taper',
        'missing_data',
      },
    );
    expect(jsonEncode(fixture), isNot(contains('@')));
    expect(jsonEncode(fixture).toLowerCase(), isNot(contains('18nburns')));
  });

  test('each timeline is chronological and carries replayable evidence', () {
    for (final timeline in timelines) {
      final days = (timeline['days'] as List).cast<Map<String, dynamic>>();
      expect(days.length, greaterThanOrEqualTo(5), reason: '${timeline['id']}');
      final dates =
          days.map((day) => DateTime.parse(day['date'] as String)).toList();
      expect(dates.toSet().length, dates.length, reason: '${timeline['id']}');
      expect([...dates]..sort(), dates, reason: '${timeline['id']}');

      for (final day in days) {
        final planned = day['planned'] as Map<String, dynamic>;
        final evidence = day['evidence'] as Map<String, dynamic>;
        final expected = day['expected'] as Map<String, dynamic>;
        expect(planned['family'], isA<String>());
        expect(planned['duration_minutes'], isA<int>());
        expect(planned['target_load'], isA<int>());
        expect(evidence['readiness'], isA<int>());
        expect(evidence['illness_symptoms'], isA<bool>());
        expect(evidence['injury_pain'], isA<bool>());
        expect(evidence['available_minutes'], isA<int>());
        expect(expected['allowed_actions'], isNotEmpty);
        expect(expected['hard_session_allowed'], isA<bool>());
      }
    }
  });

  test('missing-data timeline contains explicit null evidence', () {
    final missing = timelines.singleWhere(
      (timeline) => timeline['id'] == 'missing_data',
    );
    final days = (missing['days'] as List).cast<Map<String, dynamic>>();
    expect(
      days.any((day) =>
          (day['evidence'] as Map<String, dynamic>)['hrv_change_percent'] ==
          null),
      isTrue,
    );
  });
}
